### ISSUE TITLE
Problems with downloading and using logs from AWS

### ISSUE THREAD

**Issue by MichaelDBA** [darold/pgbadger#657]
Without customized coding, you cannot download full PG logs from AWS using the AWS CLI APIs.  Downloading files from s3 (AWS RDS/Aurora --> CloudWatch --> s3) provides another set of challenges with the weird way CloudWatch re-arranges the log data before downloading them from s3.

AWS does not download PG log files correctly using any of these APIs:
AWS CLI (truncated even with using --starting-token)
AWS SDK (the boto3 python package is broken)
S3 (log file timestamp order is not sequential)

I closed 2 other issues related to this:
issue, https://github.com/darold/pgbadger/issues/654
issue, https://github.com/darold/pgbadger/issues/656

**Comment by darold** [comment]
Are you using incremental mode? You are loading old binary files that are not compatible or are broken.

**Comment by MichaelDBA** [comment]
I am not using incremental mode.  I downloaded the latest master.zip file for pgbadger. Using 11.5 version.

**Comment by darold** [comment]
No,not necessary Perl is not the problem. I think you might probably have old binary files in your temp directory. Can you try by adding the following pgbadger option after creating directory `/tmp/issue657`:
```
pgbadger  --tempdir=/tmp/issue657  [...]
```

**Comment by MichaelDBA** [comment]
in progress, takes about 25 mins...

**Comment by darold** [comment]
If there is an error yes. 25 minutes, you are not using multiprocess?

**Comment by MichaelDBA** [comment]
ha, only got a 2 cpu EC2 instance with 4GB RAM.

So yes, got that error again.

**Comment by MichaelDBA** [comment]
No I am not using incremental mode.

**Comment by darold** [comment]
I'm a bit embarrassed with this issue, I'm not able to reproduce the problem which is difficult to be able to find the reason of the problem.

**Comment by rjuju** [comment]
> So yes, got that error again. But there is nothing in that temp file

Do you mean the `/tmp/issue657` directory or a specific temp file?  If a specific temp file, what's its name?

@darold  It could be somewhat specific to the aws platform.  Maybe for some reason Storable::store_fd / Storable::fd_retrieve isn't working as expected, or unrelated .bin files get created automatically in that directory?  Maybe some new debug messages to print all found .bin file names and the size + content of the problematic file would get a step further?

**Comment by darold** [comment]
Yes @rjuju there is probably a Storable issue but I don't think it is related to AWS as if I have well understood the files are downloaded locally before being processed by pgbadger. 

@MichaelDBA can you please execute the following command to see if it can be related to a Perl issue:
`perl -E 'use Storable qw(dclone); say "version=",$Storable::VERSION; say "limit=",$Storable::recursion_limit; my @tt; for (1..1_000_000) { my $t = [[[]]]; push @tt, $t } dclone \@tt'`

And does the host where pgbadger is run is an AWS server too and have you tried to run pgbadger  on an other host not AWS one?

**Comment by rjuju** [comment]
I understood that @MichaelDBA was running pgbadger on some EC2 instance, with files downloaded from s3.

**Comment by darold** [comment]
Ah ok, which could explain the small number of CPU, right. So let's wait for @MichaelDBA answers to my last questions, there is a know issue on Storable and Perl 5.28 fixed in Perl 5.30.

**Comment by MichaelDBA** [comment]
perl -E 'use Storable qw(dclone); say "version=",$Storable::VERSION; say "limit=",$Storable::recursion_limit; my @tt; for (1..1_000_000) { my $t = [[[]]]; push @tt, $t } dclone \@tt'
version=3.11
limit=14270
Max. recursion depth with nested structures exceeded at -e line 1.

Also I was talking about the directory you had me create.  During the run it was being populated, but afterwards it was empty.
pgbadger  --tempdir=/tmp/issue657

**Comment by darold** [comment]
Ok we've finally got it, you probably have a Perl 5.28 on this host, it has a know [Perl bug](https://github.com/perl/perl5/issues/16606) with Storable 3.x that is solved in Perl 5.30.

**Comment by MichaelDBA** [comment]
Using centOS 8, looks like I got 5.26 installed
perl.x86_64                         4:5.26.3-416.el8

**Comment by darold** [comment]
No, I have not said that with more CPU you will not have the problem, I just said that parsing will be faster. Sorry for my english.

Best is to use PerlBrew for the moment to be sure that version 5.30 will solve it. Proceed as follow:
```
sudo apt-get install perlbrew
perlbrew init
perlbrew install 5.32.1
```
then to use this version to check that pgbadger is working:
```
perlberew exec 5.32.1 /path/to/pgbadger
```
Then if the issue is gone then you can switch perl to this version by using:
```
perlbrew use 5.32.1
```
See https://docs.brew.sh/Homebrew-on-Linux#requirements for more information..

**Comment by MichaelDBA** [comment]
One last question just to make sure we are not going down the wrong path.

In most cases I do not get a parsing error (errors that occur BEFORE pgbadger starts generating the html report.
CASE 1:
-->
`Max. recursion depth with nested structures exceeded at /usr/lib64/perl5/vendor_perl/Storable.pm line 328, at /usr/local/bin/pgbadger line 15023.
`
In many cases I do not get a parsing error, but do get a pgbadger html generation error like this (CASE 2):
```
LOG: Ok, generating html report...
Use of uninitialized value in substitution iterator at /usr/local/bin/pgbadger line 19057.
2021-05-20 12:26:20  ERROR: 25
```

So is this perl fix only for case 1 above or both case 1 and case 2?

**Comment by sumeetshk** [comment]
@darold @MichaelDBA I'm also getting this error:
```
LOG: Ok, generating html report...
Use of uninitialized value in substitution iterator at /usr/bin/pgbadger line 18917.
Report could not be generated. Please check the logs.
```

AWS EC2 with 4CPU, 8GB RAM
Perl Version - 5.30 
pgBadger version 11.5
Ubuntu "20.04.2 LTS (Focal Fossa)"

Any suggestion here would be appreciated.

**Comment by darold** [comment]
Last commit 481535b might solve the use of uninitialized value in substitution iterator at line 18917, please give a try to latest development code and us me know.

**Comment by wsdookadr** [comment]
Just a thought here, it could be useful here if we could get a minimal test case for this.
A simple way to do this might be the following:
1. first determine the file that's causing this by running pgbadger individually on each of them
2. after having found a single log file that reproduces the error, it could be split into multiple chunks/parts [as described here](https://github.com/darold/pgbadger/issues/410#issuecomment-449019952) , and again checking which chunk still holds the error
3. the chunk found in step 2 can be further split again using the [same mechanism described here](https://github.com/darold/pgbadger/issues/410#issuecomment-449019952)

**Comment by darold** [comment]
I agree that a regression test must be added to test this feature.

**Comment by MichaelDBA** [comment]
I will repeat my tests trying each log individually first, but I think I already did that and the problem only occurred when I used multiple logs at one time.  Will get back to you in a couple days.

**Comment by thomasLeclaire** [comment]
Hello! 
to add some info, I have a same behavior with logs on Google cloudsql side. 
Using the last binary available on master (and perl v5.32.1, linux 5.4), I have the error : 
"Use of uninitialized value in substitution iterator at /usr/bin/pgbadger line 19133" 

I don't have this error when I run the report individually, only in incremental mode during the week report generation (the daily ones are all ok). 

I was not able to identify which part of the logs is problematic.. (and my logs are around 8G/hour).

**Comment by darold** [comment]
Commit a59ee81 might solve this issue.
